### PR TITLE
MTDSA-3083 Release 2 retrieve tax calc external audits

### PR DESCRIPTION
**Pull request by narendravijayarao**

### CODE REVIEW

File: app/v2/outcomes/TaxCalcOutcome.scala [hunk 1]
```diff
@@ -17,12 +17,14 @@
...
-import v2.models.errors.MtdError
+import v2.models.errors.{ErrorWrapper, MtdError}
 import v2.models.old.{TaxCalcMessages, TaxCalculation}
 
 object TaxCalcOutcome {
 
-  type Outcome[M] = Either[MtdError, M]
+  type Outcome[M] = Either[ErrorWrapper, DesResponse[M]]
+
+  type ServiceOutcome[M] = Either[ErrorWrapper, DesResponse[M]]
```
Comment (jeremystone) [inline app/v2/outcomes/TaxCalcOutcome.scala]: Never used

File: app/v2/outcomes/TaxCalcOutcome.scala [hunk 2]
```diff
@@ -17,12 +17,14 @@
...
 
   type TaxCalcOutcome = Either[MtdError, TaxCalculation]
```
Comment (jeremystone) [inline app/v2/outcomes/TaxCalcOutcome.scala]: Never used

File: app/v2/outcomes/TaxCalcOutcome.scala [hunk 3]
```diff
@@ -17,12 +17,14 @@
...
   type TaxCalcMessagesOutcome = Either[MtdError, TaxCalcMessages]
```
Comment (jeremystone) [inline app/v2/outcomes/TaxCalcOutcome.scala]: Never used

File: app/v2/outcomes/TaxCalcOutcome.scala (hunk 2 above)
Comment (narendravijayarao) [reply app/v2/outcomes/TaxCalcOutcome.scala]: It is used in some tests, don't know why!! will delete if it is not used anywhere else

File: test/v2/controllers/TaxCalcControllerSpec.scala [hunk 4]
```diff
@@ -96,27 +115,42 @@ class TaxCalcControllerSpec extends ControllerBaseSpec {
...
+        val event = AuditEvent("retrieveTaxCalculation",
+          "mtd-tax-calculation", detail)
+        MockedAuditService.verifyAuditEvent(event).once
       }
 
       "the service returns an Invalid CalcID error" in new Test {
         MockedMtdIdLookupService.lookup(nino)
           .returns(Future.successful(Right(mtdId)))
-
+        MockedEnrolmentsAuthService.authoriseUser().returns(Future.successful(expected))
```
Comment (jeremystone) [inline test/v2/controllers/TaxCalcControllerSpec.scala]: Were we not authorising before this change under some error scenarios?

File: app/v2/controllers/TaxCalcController.scala [hunk 5]
```diff
@@ -16,59 +16,106 @@
...
+      case Right(v2.outcomes.DesResponse(correlationId, responseData)) =>
+        if (featureSwitch.isRelease2Enabled && responseData.isInstanceOf[TaxCalculation]) {
+          auditSubmission(RetrieveTaxCalcAuditDetail(userDetails.userType, userDetails.agentReferenceNumber,
+            nino, calcId, correlationId, RetrieveTaxCalcAuditResponse(OK, None, Some(toJson(responseData)))))
+        }
+        Ok(toJson(responseData))
+      case Left(errorWrapper) =>
+        val correlationId = getCorrelationId(errorWrapper)
+        val result = processError(errorWrapper).withHeaders("X-CorrelationId" -> correlationId)
+        if (featureSwitch.isRelease2Enabled) {
```
Comment (jeremystone) [inline app/v2/controllers/TaxCalcController.scala]: This will audit TaxCalcMessages in error scenarios. Not sure this is the intended behaviour.

Why not have a boolean parameter 'audit' for this 'get' method and use this here and at line 76 to decide whether to audit (which also has the benefit of not having the isInstanceOf)..?

File: app/v2/controllers/TaxCalcController.scala [hunk 6]
```diff
@@ -16,59 +16,106 @@
...
+    f(nino,calcId).map {
+      case Right(v2.outcomes.DesResponse(correlationId, responseData)) =>
+        if (featureSwitch.isRelease2Enabled && responseData.isInstanceOf[TaxCalculation]) {
+          auditSubmission(RetrieveTaxCalcAuditDetail(userDetails.userType, userDetails.agentReferenceNumber,
+            nino, calcId, correlationId, RetrieveTaxCalcAuditResponse(OK, None, Some(toJson(responseData)))))
+        }
+        Ok(toJson(responseData))
+      case Left(errorWrapper) =>
+        val correlationId = getCorrelationId(errorWrapper)
+        val result = processError(errorWrapper).withHeaders("X-CorrelationId" -> correlationId)
```
Comment (jeremystone) [inline app/v2/controllers/TaxCalcController.scala]: We're not adding the X-CorrelationId header for success case and we weren't adding it before this change. Should doing so be a part of another story?

File: app/v2/models/errors/desError.scala [hunk 7]
```diff
@@ -17,30 +17,14 @@
 package v2.models.errors
 
 import play.api.libs.json._
-import v2.models.errors.DesErrorCode.DesErrorCode
 
-object DesErrorCode extends Enumeration {
```
Comment (jeremystone) [inline app/v2/models/errors/desError.scala]: Why the change? I'm not a fan of Enumeration either but at least it's type safe. (I realise we use strings elsewhere for these error codes but we use a map in a single place to convert between them and the MTD errors).

Comment (narendravijayarao) [reply app/v2/models/errors/desError.scala]: The way we implement in this one is different, we will do all changes step by step

File: test/v2/controllers/TaxCalcControllerSpec.scala (hunk 4 above)
Comment (narendravijayarao) [reply test/v2/controllers/TaxCalcControllerSpec.scala]: what?? We are using Auth before also

File: app/v2/controllers/TaxCalcController.scala (hunk 6 above)
Comment (narendravijayarao) [reply app/v2/controllers/TaxCalcController.scala]: We should...

File: test/v2/controllers/TaxCalcControllerSpec.scala (hunk 4 above)
Comment (jeremystone) [reply test/v2/controllers/TaxCalcControllerSpec.scala]: What I mean is: why the new expectation (test now fails without it); but the test must have passed without the expectation with the old code. So are we doing more auth than before? 

If as you say we're not, then I guess it's ok.
